constraint-solver/src/indexed_constraint_system.rs

```diff
@@ -280,6 +281,68 @@ impl<T: RuntimeConstant + Substitutable<V>, V: Clone + Hash + Ord + Eq>
     }
 }
 
+impl<T: RuntimeConstant + Substitutable<V> + Hash, V: Clone + Hash + Ord + Eq>
+    IndexedConstraintSystem<T, V>
+{
+    pub fn apply_expression_substitutions(
+        &mut self,
+        substitutions: impl IntoIterator<Item = (GroupedExpression<T, V>, GroupedExpression<T, V>)>,
+    ) {
+        for (expression, substitution) in substitutions {
+            if let Some(var) = expression.try_to_simple_unknown() {
+                self.substitute_by_unknown(&var, &substitution);
+            } else {
+                self.substitute_expression_in_bus_interactions(&expression, &substitution);
+            }
+        }
+    }
+}
+
+impl<T: RuntimeConstant + Hash, V: Clone + Hash + Ord + Eq> IndexedConstraintSystem<T, V> {
+    pub fn substitute_expression_in_bus_interactions(
```

Comment:
I think this could be private. Also, maybe add a docstring explaining that it only replaces entire bus interaction fields, i.e., it doesn't replace any subexpressions.